src/lock.rs

```diff
@@ -0,0 +1,166 @@
+// (c) Copyright 2023 Helsing GmbH. All rights reserved.
+
+use std::{collections::HashMap, sync::Arc};
+
+use eyre::{ensure, Context};
+use ring::digest;
+use semver::{Version, VersionReq};
+use serde::{Deserialize, Serialize};
+use tokio::fs;
+
+use crate::{
+    manifest::{Dependency, DependencyManifest},
+    package::{Package, PackageName},
+};
+
+pub const LOCKFILE: &str = "Proto.lock";
+
+#[derive(Serialize, Deserialize, Clone)]
+pub struct LockedPackage {
+    pub name: PackageName,
+    pub checksum: String,
+    pub repository: String,
+    pub version: Version,
+    pub dependencies: Vec<PackageName>,
+}
+
+impl LockedPackage {
+    pub fn lock(package: &Package, repository: String) -> Self {
+        let checksum = digest::digest(&digest::SHA256, &package.tgz);
+
+        Self {
+            name: package.name().to_owned(),
+            repository,
+            checksum: hex::encode(checksum),
+            version: package.version().to_owned(),
+            dependencies: package
+                .manifest
+                .dependencies
+                .iter()
+                .cloned()
+                .map(|d| d.package)
+                .collect(),
+        }
+    }
+
+    pub fn validate(&self, other: &Self) -> eyre::Result<()> {
```

Comment:
Can you add doc comments to everything that is public?

Reply:
Think I covered everything now (🤞 ).